### ISSUE TITLE
ConfItem alias circularly references itself

### ISSUE THREAD

**Issue by mikerockett** [tmont/nginx-conf#29]
> node_modules/nginx-conf/src/conf.d.ts:44:21 - error TS2456: Type alias 'ConfItem' circularly references itself.
> 44 export declare type ConfItem = ConfItemContext & {

New to the world of TS, so not 100% sure if this is because of the way in which I'm using nginx-conf, but my gut tells me this is not the case. (Let me know if you'd like to see my implementation, please.)

For reference, I'm using this in an [oclif](https://github.com/oclif/oclif) app, and it works fine when I run it directly (using `./bin/run`). However, I get the above when I try to publish my CLI app, which runs `tsc -b` before packing the tarball.

Thanks!

**Comment by mikerockett** [comment]
As an aside, I'm also getting the below error, in the same app, and also only on publish:

> node_modules/nginx-conf/src/conf.d.ts:1:30 - error TS2307: Cannot find module 'tiny-typed-emitter' or its corresponding type declarations.

**Comment by tmont** [comment]
Hmmm. I'm just gonna remove the TypeScript stuff for now. I guess you need `tiny-typed-emitter` as a peer dependency? I thought that looked suspect when I merged it.

**Comment by tmont** [comment]
Janky TypeScript has been removed in version `1.7.0`. Sorry about that!

**Comment by mikerockett** [comment]
>  I guess you need tiny-typed-emitter as a peer dependency?

I don't need it for anything – it just moaned about it. Had a similar issue with listr2 depending on enquirer – landed up just requiring it and enabling ES module interop. 🤷🏻‍♂️

**Comment by jirutka** [comment]
@mikerockett, you have too old typescript version. IIRC the types require 4.1.

@tmont, why the heck did you completely remove the type declarations?! You could at least ping me, I’d help. Removing types is not a backward compatible change, you broke the package for everyone who have started using them. Including myself, I use them in package [nginx-testing](https://www.npmjs.com/package/nginx-testing).

**Comment by jirutka** [comment]
> I guess you need tiny-typed-emitter as a peer dependency? I thought that looked suspect when I merged it.

No, this problem has nothing to do with `tiny-typed-emitter`. And the only reason why I’ve added `tiny-typed-emitter` as a peer dependency was to avoid pushing TS-only dependency to _all_ users of nginx-conf.

**Comment by mikerockett** [comment]
> @mikerockett, you have too old typescript version. IIRC the types require 4.1.

You are correct – didn't realise. I can see how it was easy to miss though, as global TS is not checked at install-time. That aside, with the amount of TS issues I bumped into since this issue was opened, I've landed up rewriting the app in V. It's just a CLI app at the end of the day, so seems apt.

**Comment by tmont** [comment]
Sorry for the confusion everyone.

I was actually in the process of rewriting this in TypeScript, anyway, so that's why I pulled the trigger on the previous declarations so quick (since I was going to delete them soon anyway).

I've just published `v2.0.0` to NPM, which has a new (breaking!) interface. In addition to that major change, I also rewrote everything in TypeScript and the emitted declarations are part of the NPM package (in addition to the source maps, should you need those).

You can view the README as I added some more detail there, but pretty much everything is the same except you need to write more angle brackets:

```typescript
// before:
conf.nginx.http.server._value

// after:
conf.nginx.http[0].server[0]._value

// in typescript you can use the null-coalesce operator to not have to use type guards everywhere
conf.nginx.http?.[0].server?.[0]._value

// or, using type guards
if (conf.nginx.http && conf.nginx.http[0].server) {
  conf.nginx.http[0].server[0]._value
}
```

@jirutka I didn't mean to trample on your contribution, but I didn't want to have a breaking change that I didn't understand out on NPM. Since at least one person went to the trouble to create an issue indicating they were having compilation issues after a clean install, I decided to play it safe and revert the change.

I incremented the minor version so if you were using something like `1.6.x` or pinning to the specific version in your package.json it shouldn't have automatically picked it up.

----

Thanks for the reports, and again, sorry for confusion!
